src/axom/core/ArrayBase.hpp

```diff
@@ -1261,14 +1350,26 @@ struct ArrayOpsBase<T, true>
    */
   static void move(T* array, IndexType src_begin, IndexType src_end, IndexType dst)
   {
-    // Since this memory is on the device-side, we copy it to a temporary buffer
-    // first.
-    IndexType nelems = src_end - src_begin;
-    T* tmp_buf =
-      axom::allocate<T>(nelems, axom::execution_space<ExecSpace>::allocatorID());
-    axom::copy(tmp_buf, array + src_begin, nelems * sizeof(T));
-    axom::copy(array + dst, tmp_buf, nelems * sizeof(T));
-    axom::deallocate(tmp_buf);
+    if(!std::is_trivially_copyable<T>::value &&
+       SPACE == OperationSpace::Unified_Device)
+    {
+      // Type might not be trivially-relocatable, move the range on the host.
+      // Note that we only do this for objects in unified/pinned memory, since
+      // we assume that objects in device-only memory are trivially-relocatable.
+      HostOp::move(array, src_begin, src_end, dst);
+    }
+    else
+    {
+      // Since this memory is on the device-side, we copy it to a temporary buffer
+      // first.
```

Comment:
Out of curiosity, why do we have to copy to a temporary buffer first?  Isn't `axom::copy` smart enough to do the right thing based on where src and dst are?

Reply:
My recollection is that there was some issue relating to memory overlap handling that required us to handle this case manually. But I could be wrong -- I'll take a closer look at this.